warehouse/query-core/src/main/java/datawave/query/jexl/DelayedIndexOnlyContext.java

```diff
@@ -0,0 +1,131 @@
+    private DatawaveJexlContext delegate;
+    private Set<String> delayedFields;
+    private IteratorBuildingVisitor iteratorBuildingVisitor;
+    private Set<JexlNode> delayedSubTrees;
+    private Range docRange;
+    private Equality equality;
+    
+    /**
+     * track which fields have been fetched already
+     */
+    private Set<String> fetched;
+    
+    public DelayedIndexOnlyContext(DatawaveJexlContext delegate, IteratorBuildingVisitor iteratorBuildingVisitor, Set<JexlNode> delayedSubTrees,
+                    Set<String> delayedFields, Range docRange, Equality equality) {
+        this.delegate = delegate;
+        this.iteratorBuildingVisitor = iteratorBuildingVisitor;
+        this.delayedSubTrees = delayedSubTrees;
+        this.delayedFields = delayedFields;
+        this.docRange = docRange;
+        this.equality = equality;
+        
+        fetched = new HashSet<>();
+    }
+    
+    @Override
+    public void set(String name, Object value) {
+        delegate.set(name, value);
+    }
+    
+    @Override
+    public Object get(String name) {
+        // only do something special if there is delayed work to do
+        if ((null != name) && delayedFields.contains(name) && !fetched.contains(name)) {
+            // fetch the field that was delayed
+            List<Document> documentFragments = null;
+            try {
+                documentFragments = fetchOnDemand(name);
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to fetch delayed index only fragments for field: " + name, e);
+            }
+            
+            // aggregate all of that into the delegate
+            for (Document document : documentFragments) {
+                document.visit(Collections.singleton(name), delegate);
+            }
+            
+            // flag that this field has now been fetched
+            fetched.add(name);
+        }
+        
+        return delegate.get(name);
+    }
+    
+    /**
+     * Use the IteratorBuildingVisitor limit to the current docRange to parse all delayed sub trees of the query. From those delayed sub trees initialize all
+     * iterators matching the target field and aggregate all partial Documents into a list
+     * 
+     * @param name
+     *            the name of the field to fetch on demand
+     * @return the list of Document objects that were fetched from all delayed iterators associated with the on-demand field
+     */
+    private List<Document> fetchOnDemand(String name) throws IOException {
+        List<Document> documentList = new ArrayList<>();
+        
+        // limit the ranges to use to the current document
+        iteratorBuildingVisitor.limit(docRange);
+        
+        // for each sub tree build the nested iterator
+        for (JexlNode subTree : delayedSubTrees) {
+            // short circuit trees that have no identifiers
+            if (JexlASTHelper.getIdentifiers(subTree).size() == 0) {
```

Comment:
This does a full scan of the underlying tree.  Is that worth it?

Reply:
This case is actually covered in the DelayedNonEventSubTreeVisitor now, so will remove